### ISSUE TITLE
Issues when setting gamma grid

### ISSUE THREAD

**Issue by mwaskom** [psychopy/psychopy#1202]
Hi,

I'm getting a crash when trying to set the gamma information for my monitor. Here is a minimal script that can cause the crash:

```python
import numpy as np
from psychopy import visual
from psychopy.monitors import Monitor

m = Monitor(name="test",
            width=40,
            distance=52,
            gamma=2.2,
            )

gamma_grid=np.array([[.001, 92.9, 2.21],
                     [.007, 18.5, 2.24],
                     [-.016, 64.3, 2.22],
                     [-.019, 10.5, 2.22]])
m.setGammaGrid(gamma_grid)

win = visual.Window(monitor=m)
```

And here is the stack trace:

```
/Users/mwaskom/anaconda/envs/psychopy/lib/python2.7/site-packages/psychopy/visual/helpers.py:173: FutureWarning: comparison to `None` will result in an elementwise object comparison in the future.
  if color==None:
/Users/mwaskom/anaconda/envs/psychopy/lib/python2.7/site-packages/psychopy/monitors/calibTools.py:196: FutureWarning: comparison to `None` will result in an elementwise object comparison in the future.
  if thisGamma == None \
Traceback (most recent call last):
  File "psychopy_gamma_bug.py", line 17, in <module>
    win = visual.Window(monitor=m)
  File "/Users/mwaskom/anaconda/envs/psychopy/lib/python2.7/site-packages/psychopy/visual/window.py", line 324, in __init__
    self._setupGamma()
  File "/Users/mwaskom/anaconda/envs/psychopy/lib/python2.7/site-packages/psychopy/visual/window.py", line 1079, in _setupGamma
    self.setGamma(self.gamma)  # using either pygame or bits++
  File "/Users/mwaskom/anaconda/envs/psychopy/lib/python2.7/site-packages/psychopy/visual/window.py", line 1099, in setGamma
    self.winHandle.setGamma(self.winHandle, self.gamma)
  File "/Users/mwaskom/anaconda/envs/psychopy/lib/python2.7/site-packages/psychopy/gamma.py", line 37, in setGamma
    setGammaRamp(pygletWindow, newLUT)
  File "/Users/mwaskom/anaconda/envs/psychopy/lib/python2.7/site-packages/psychopy/gamma.py", line 59, in setGammaRamp
    error =carbon.CGSetDisplayTransferByTable(pygletWindow._screen.id, LUTlength,
AttributeError: 'CocoaScreen' object has no attribute 'id'
```

System information:

OS: Mac OSX 10.9.5
Psychopy version: 1.80.01dev

I am basically using the versions of things that are installed through conda when doing `conda -c erik install psychopy`. Here is a full list of what I have installed (i.e. the output of `conda list`):

```
# packages in environment at /Users/mwaskom/anaconda/envs/psychopy:
#
altgraph                  0.11                     py27_0    https://conda.binstar.org/erik/osx-64/altgraph-0.11-py27_0.tar.bz2
appnope                   0.1.0                    py27_0    defaults
backports                 1.0                      py27_0    defaults
backports-abc             0.4                       <pip>
backports.shutil-get-terminal-size 1.0.0                     <pip>
backports.ssl-match-hostname 3.4.0.2                   <pip>
backports_abc             0.4                      py27_0    defaults
bdist-mpkg                0.5.0                     <pip>
bdist_mpkg                0.5.0                    py27_0    https://conda.binstar.org/erik/osx-64/bdist_mpkg-0.5.0-py27_0.tar.bz2
certifi                   14.05.14                 py27_0    <unknown>
cffi                      1.1.2                    py27_0    defaults
colormath                 2.1.1                     <pip>
configparser              3.5.0b2                  py27_1    defaults
cregg (/Users/mwaskom/Code/cregg) 0.1.dev0                  <pip>
cython                    0.23.2                   py27_0    defaults
dateutil                  2.4.1                    py27_0    defaults
decorator                 4.0.10                   py27_0    defaults
docutils                  0.12                     py27_0    <unknown>
entrypoints               0.2.2                    py27_0    defaults
freetype                  2.5.5                         1    defaults
functools32               3.2.3.2                  py27_0    defaults
get_terminal_size         1.0.0                    py27_0    defaults
gevent                    1.0.1                    py27_0    <unknown>
greenlet                  0.4.9                    py27_0    defaults
hdf5                      1.8.9                         2    defaults
iolabs                    3.2                      py27_0    https://conda.binstar.org/erik/osx-64/iolabs-3.2-py27_0.tar.bz2
ipykernel                 4.3.1                    py27_0    defaults
ipython                   4.2.0                    py27_1    defaults
ipython-genutils          0.1.0                     <pip>
ipython-notebook          4.0.4                    py27_0    defaults
ipython_genutils          0.1.0                    py27_0    defaults
ipywidgets                4.1.1                    py27_0    defaults
jdcal                     1.0                      py27_0    <unknown>
jinja2                    2.8                      py27_1    defaults
jpeg                      8d                            1    <unknown>
jsonschema                2.5.1                    py27_0    defaults
jupyter                   1.0.0                    py27_3    defaults
jupyter-client            4.3.0                     <pip>
jupyter-console           4.1.1                     <pip>
jupyter-core              4.1.0                     <pip>
jupyter_client            4.3.0                    py27_0    defaults
jupyter_console           4.1.1                    py27_0    defaults
jupyter_core              4.1.0                    py27_0    defaults
labjackpython             10222012                 py27_0    https://conda.binstar.org/erik/osx-64/labjackpython-10222012-py27_0.tar.bz2
lcms                      1.19                          0    defaults
libpng                    1.6.22                        0    defaults
libsodium                 0.4.5                         0    <unknown>
libxml2                   2.9.2                         0    <unknown>
libxslt                   1.1.28                        2    <unknown>
lxml                      3.4.4                    py27_0    <unknown>
macholib                  1.6                      py27_0    https://conda.binstar.org/erik/osx-64/macholib-1.6-py27_0.tar.bz2
markupsafe                0.23                     py27_2    defaults
matplotlib                1.4.3                np19py27_3    defaults
mistune                   0.7.2                    py27_1    defaults
modulegraph               0.11.1                   py27_0    https://conda.binstar.org/erik/osx-64/modulegraph-0.11.1-py27_0.tar.bz2
nbconvert                 4.2.0                    py27_0    defaults
nbformat                  4.0.1                    py27_0    defaults
networkx                  1.10                      <pip>
nose                      1.3.7                    py27_0    <unknown>
notebook                  4.2.1                    py27_0    defaults
numexpr                   2.3.1                np18py27_0    defaults
numpy                     1.10.1                   py27_0    defaults
openpyxl                  2.0.2                    py27_0    defaults
openssl                   1.0.2h                        1    defaults
pandas                    0.17.0              np110py27_0    defaults
path.py                   8.2.1                    py27_0    defaults
pathlib2                  2.1.0                    py27_0    defaults
patsy                     0.4.0               np110py27_0    defaults
pep8                      1.6.2                     <pip>
pexpect                   4.0.1                    py27_0    defaults
pickleshare               0.7.2                    py27_0    defaults
pil                       1.1.7                    py27_1    defaults
pip                       8.1.2                    py27_0    defaults
portaudio                 19                            0    defaults
psutil                    3.2.1                    py27_0    defaults
psychopy                  1.80.01dev           np18py27_1    https://conda.binstar.org/erik/osx-64/psychopy-1.80.01dev-np18py27_1.tar.bz2
ptyprocess                0.5.1                    py27_0    defaults
py2app                    0.8                      py27_0    https://conda.binstar.org/erik/osx-64/py2app-0.8-py27_0.tar.bz2
pycparser                 2.14                     py27_0    <unknown>
pyglet                    1.2a1                     <pip>
pyglet1.2                 1.2alpha1                py27_1    https://conda.binstar.org/erik/osx-64/pyglet1.2-1.2alpha1-py27_1.tar.bz2
pygments                  2.1.3                    py27_0    defaults
pyobjc-core               2.5.1                    py27_0    https://conda.binstar.org/erik/osx-64/pyobjc-core-2.5.1-py27_0.tar.bz2
pyobjc-framework-cocoa    2.5.1                    py27_0    https://conda.binstar.org/erik/osx-64/pyobjc-framework-cocoa-2.5.1-py27_0.tar.bz2
pyobjc-framework-qtkit    2.5.1                    py27_0    https://conda.binstar.org/erik/osx-64/pyobjc-framework-qtkit-2.5.1-py27_0.tar.bz2
pyobjc-framework-quartz   2.5.1                    py27_0    https://conda.binstar.org/erik/osx-64/pyobjc-framework-quartz-2.5.1-py27_0.tar.bz2
pyopengl                  3.0.2                    py27_0    defaults
pyparsing                 2.0.3                    py27_0    <unknown>
pyqt                      4.11.4                   py27_3    defaults
pyserial                  2.7                      py27_0    https://conda.binstar.org/erik/osx-64/pyserial-2.7-py27_0.tar.bz2
pysoundcard               0.5.0                np18py27_0    https://conda.binstar.org/erik/osx-64/pysoundcard-0.5.0-np18py27_0.tar.bz2
pytables                  3.1.1                np18py27_0    defaults
python                    2.7.12                        0    defaults
python-dateutil           2.4.2                    py27_0    <unknown>
python.app                1.2                      py27_4    <unknown>
pytz                      2015.7                   py27_0    defaults
pyxid                     1.0                      py27_0    https://conda.binstar.org/erik/osx-64/pyxid-1.0-py27_0.tar.bz2
pyyaml                    3.11                     py27_1    <unknown>
pyzmq                     15.2.0                   py27_1    defaults
qt                        4.8.7                         3    defaults
qtconsole                 4.2.1                    py27_0    defaults
readline                  6.2                           2    <unknown>
scipy                     0.16.0              np110py27_1    defaults
seaborn                   0.6.0                np19py27_0    defaults
setuptools                23.0.0                   py27_0    defaults
simplegeneric             0.8.1                    py27_1    defaults
singledispatch            3.4.0.3                  py27_0    defaults
sip                       4.16.9                   py27_0    defaults
six                       1.10.0                   py27_0    defaults
sphinx                    1.2.3                    py27_0    defaults
sqlite                    3.13.0                        0    defaults
ssl_match_hostname        3.4.0.2                  py27_1    defaults
statsmodels               0.6.1               np110py27_0    defaults
tables                    3.1.1                     <pip>
terminado                 0.6                      py27_0    defaults
tk                        8.5.18                        0    <unknown>
tornado                   4.3                      py27_1    defaults
traitlets                 4.2.1                    py27_0    defaults
wheel                     0.29.0                   py27_0    defaults
wxpython                  3.0                      py27_0    defaults
wxpython-common           3.0.0.0                   <pip>
yaml                      0.1.6                         0    <unknown>
zeromq                    4.1.3                         0    defaults
zlib                      1.2.8                         3    defaults
```

**Comment by mwaskom** [comment]
Ping @peircej , sorry to be annoying but this is a rather debilitating issue if I want to do psychophysics.

**Comment by mwaskom** [comment]
Also for the record I had to upgrade to Psychopy 1.81.03 to get iohub working and am still seeing the issue in that environment.

**Comment by mwaskom** [comment]
I am almost completely ignorant about the OSX GUI layer, but I have tried to debug this a bit myself. I think this may be fixed in Psychopy 1.84, specifically with this line: https://github.com/psychopy/psychopy/blob/master/psychopy/visual/gamma.py#L85

I am trying to figure out how to install 1.84 to see if this lets my code run.

However one thing I am still confused about is that even in psychopy 1.84, `gamma.py` is still loading the Carbon library, but my `Window.winHandle` is a `<pyglet.window.cocoa.CocoaWindow at 0x1046228d0>`. Is this going to be a problem? With a quick skim of `visual/window.py` it's not clear how this would be configured.

**Comment by mwaskom** [comment]
As an update, if I make psychopy actually do something (specifically, the [drifting plaid demo](http://www.psychopy.org/demos/plaid.html)), in this script, I find that the gamma table I have set is in effect only for a brief time at the beginning of the script, and then resets. It seems to reset after 5-30 window flips, with the actual timing appearing to be random.

**Comment by peircej** [comment]
Yeah, I'm sure that was fixed in Jan so you need to update
Your options are:

1. download a new standalone for 1.84 (but that's 64bit and still a bit of a work in progress)
2. just download the zip release for 1.83.04 and combine that with your existing (32bit) standalone:

  - you'd want this file: https://github.com/psychopy/psychopy/releases/download/1.83.04/PsychoPy-1.83.04.zip
  - Then go to menu tools>update and browse for the zip file you downloaded (hopefully mac didn't unzip it automatically)

Make sense? If you update to 1.83.04 or 1.84.0 and still have a problem then I'll look at it further but I'm not going to downgrade my own version to dig around for that problem

cheers, 
Jon

**Comment by mwaskom** [comment]
Just to be clear, by "that problem" do you mean psychopy crashing when setting the gamma table, or the gamma table resetting?

**Comment by mwaskom** [comment]
Apologies that [this comment](https://github.com/psychopy/psychopy/issues/1202#issuecomment-230809760) was not fully clear, but I had installed 1.83 (by doing `pip install https://github.com/psychopy/psychopy/releases/download/1.83.04/PsychoPy-1.83.04.zip`) and was able to move past the crash, but I have encountered another problem as described above.

**Comment by peircej** [comment]
Ah. So you did update to newer version.

Well, I just tried to combine your gamma snippet from above with the code from demos>stimuli>gabor.py
```
#!/usr/bin/env python2
# -*- coding: utf-8 -*-


from psychopy import core, visual, event
from psychopy.monitors import Monitor
import numpy as np

m = Monitor(name="test",
            width=40,
            distance=52,
            gamma=2.2,
            )

gamma_grid=np.array([[.001, 92.9, 2.21],
                     [.007, 18.5, 2.24],
                     [-.016, 64.3, 2.22],
                     [-.019, 10.5, 2.22]])
m.setGammaGrid(gamma_grid)

# create a window to draw in
win = visual.Window([400, 400.0], allowGUI=False, monitor=m)

# INITIALISE SOME STIMULI
gabor = visual.GratingStim(win, tex="sin", mask="gauss", texRes=256, 
           size=[1.0, 1.0], sf=[4, 0], ori = 0, name='gabor1')
gabor.autoDraw = True
message = visual.TextStim(win, pos=(0.0, -0.9), text='Hit Q to quit')
trialClock = core.Clock()

# repeat drawing for each frame
while trialClock.getTime() < 20: 
    gabor.phase += 0.01
    message.draw()
    # handle key presses each frame
    if event.getKeys(keyList=['escape', 'q']):
        win.close()
        core.quit()

    win.flip()

win.close()
core.quit()


```

The code above gave me no problems; it didn't revert the gamma.

Don't know why you're having problems. Could you run that exact code? Could you try on Standalone distribution rather than using anaconda?

**Comment by peircej** [comment]
The cocoa/carbon thing is a red herring, by the way. It is using the framework CoreGraphics where that's found and reverts to Carbon where not, as determined by the lines at the top of gamma.py

**Comment by mwaskom** [comment]
I ran your script and am seeing the same problem where the gamma reverts after ~20 frames.

I downloaded the "standalone version" and installed it to `/Applications`, but if you'll forgive the dumb question, I can't figure out how to execute your script with this version of PsychoPy. Doing `/Applications/PsychoPy2.app/Contents/MacOS/python jp_script.py` complains that

```
Could not find platform independent libraries <prefix>
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
ImportError: No module named site
```

I can make the PsychoPy application open, but I can't find anything in the docs about how to use the Python binary in your standalone distribution to simply execute a python script?

**Comment by mwaskom** [comment]
I *think* I managed to open the script in the "PsychoPy Coder" and ran it by clicking the little green runner button and am seeing the same issue — the gamma of the whole screen briefly changes while the  experiment runs and flips a few frames, but then quickly reverts while the experiment continues on with no other sign of problems.

**Comment by peircej** [comment]
You open the app. From the app you open the file. Then you press the run icon


This was sent from my phone. Apologies for brevity, typos and inappropriate auto-corrects



-------- Original message --------

**Comment by mwaskom** [comment]
Yes please see the second comment...

**Comment by peircej** [comment]
As I say, I can't reproduce that problem. On my machines the gamma stays until the screen closes, and I have no idea why it would be different on your computer. I'm certain psychopy doesn't change gamma back to some other value (unless you explicitly do so). Maybe you have some other software running that is doing it? But I'm pretty sure this isn't a bug in PsychoPy so closing this for now

Sorry I couldn't help

**Comment by mwaskom** [comment]
Great, thanks for all your help.

**Comment by DRezeanu** [comment]
Hey guys, hate to open a VERY old thread, but I'm having this same issue. I've managed to work around several problems with the latest build, but no matter how I work around it, any gamma setting disappears within a few frame flips. This applies no matter how I set the gamma: monitor center, gammagrid, or even manually pulling the screenID and calling SetDeviceGammaRamp through ctypes. 

I realize this is probably beyond the scope of this old thread, and I'm planning to post in a more visible place to see if anyone can help, but the inability to properly linearize the display is basically a no-go for the experiments we're running.

Thanks in advance if you guys can help!
